plat/common/x86/syscall.S

```diff
@@ -65,48 +67,96 @@ ENTRY(_ukplat_syscall)
+	 *  ukarch_execenv| |-------------| v             |
+	 *                | |             | ^             |
+	 *                | |  struct     | |             |
+	 *                | | ukarch_ectx | |ECTX_SIZE    |
+	 *                | |             | |             |
+	 *                | |             | |             |
+	 *                | |             | |             |
+	 *                v +-------------+ v             v
+	 *                       stack
+	 *                       pointer
+	 *
+	 * Where ukarch_sysctx/ukarch_ectx/__regs we fill in below.
+	 */
+	subq	$(UKARCH_EXECENV_END_ALIGN - 8), %rsp
+	.cfi_adjust_cfa_offset (UKARCH_EXECENV_END_ALIGN - 8)
+
+	pushq	%gs:LCPU_AUXSP_OFFSET
+	.cfi_adjust_cfa_offset 8
+
+	/**
+	 * We are done getting what we needed from KERNEL_GS_BASE, swap back.
+	 * We do this immediately so that we avoid confusions like:
+	 * - I am in an exception handler, was I in a syscall or application
+	 * code when it happened? What is the value of KERNEL_GS_BASE/GS_BASE
+	 * - I am returning from clone(), should I do swapgs? Was this an
+	 * internal clone() call?
+	 * ...
+	 *
+	 * By doing this we will always know:
+	 * - Were we interrupted/trapped from Unikraft/syscall code? Then
+	 * GS_BASE == lcpu
+	 * - Were we interrupted/trapped from application code? Then we don't
+	 * know GS_BASE so we need to get lcpu from KERNEL_GS_BASE.
+	 */
+	swapgs
+
 	/* NOTE: We should normally align the stack before doing this
 	 * subtraction because we must ensure that the `ectx` field
 	 * is aligned to the corresponding ECTX alignment.
 	 * However, this is guaranteed to already be the case for the
 	 * auxiliary stack because it is allocated with this exact alignment
 	 * in mind.
 	 */
-	subq	$(UK_SYSCALL_CTX_SIZE - __REGS_SIZEOF), %rsp
-	.cfi_adjust_cfa_offset (UK_SYSCALL_CTX_SIZE - __REGS_SIZEOF)
+	subq	$(UKARCH_EXECENV_SIZE - __REGS_SIZEOF), %rsp
+	.cfi_adjust_cfa_offset (UKARCH_EXECENV_SIZE - __REGS_SIZEOF)
 
 	pushq_cfi	$(GDT_DESC_OFFSET(GDT_DESC_DATA))
 
-	/* Store application's stack pointer at the top of current thread's
-	 * auxiliary stack. We have to do this because we obviously can't
-	 * rely on the scratch register being maintained between thread switches
-	 */
 	pushq_reg_cfi r11
 	.cfi_rel_offset rsp, 0
 
-	/* We are now in a state where the stack looks like this:
-	 *	--------------- <-- auxsp (i.e. lcpu_get_current()->auxsp,
-	 *	| app's saved |	           i.e. uk_thread_current()->auxsp)
-	 *	|     %ss     |
-	 *	---------------
-	 *	| app's saved |
-	 *	|     %rsp    |
-	 *	--------------- <-- (auxsp - 16) i.e. (**current %rsp**)
-	 *	|             |
-	 *	|             |
-	 *            ...
-	 *	|             |
-	 *	--------------- <-- (auxsp -
-	 *			     (PAGE_SIZE *
-	 *			      (1 << CONFIG_UKPLAT_AUXSP_PAGE_ORDER)))
-	 *	  END OF AUXSP
-	 */
-
 	/*
 	 * Push arguments in the order of 'struct __regs' to the stack.
 	 * We are going to handover a refernce to this stack area as
```

Comment:
TODO: set IRQ enable flag in saved eflags so that stored ctx is reflects the reality more honestly - an application will never disable IRQ's.